File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Location.java
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.Command;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;

import java.util.*;

public class Location extends ICommand {

    public Location(String name) {
        super(name);
    }

    @Override
    public void onPlayerCommand(Player sp, Command cmd, String label, String[] args) {
        if (TosoGameAPI.isAdmin(sp)) {
            WorldConfig worldConfig = Main.getWorldConfig();

            if (args.length != 0) {
                if (args[0].equalsIgnoreCase("opgame") || args[0].equalsIgnoreCase("opg")) {
                    worldConfig.getOPGameLocationConfig().setOPLocation(sp.getLocation());
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "オープニングゲーム地点を設定しました。");
                    return;
                } else if (args[0].equalsIgnoreCase("gopgame") || args[0].equalsIgnoreCase("gopg")) {
                    if (args.length != 1) {
                        try {
                            int i = Integer.parseInt(args[1]);
                            if (i >= 1) {
                                worldConfig.getOPGameLocationConfig().setGOPLocation(i, sp.getLocation());
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "オープニングゲーム集合地点の位置" + i + "を設定しました。");
                                return;
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        } catch (NumberFormatException e) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        }
                    }
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location " + args[0] + " <数値>\" - オープニングゲーム集合地点の設定");
                    return;
                } else if (args[0].equalsIgnoreCase("hunter")) {
                    if (args.length != 1) {
                        try {
                            int i = Integer.parseInt(args[1]);
                            if (i >= 1) {
                                worldConfig.getHunterLocationConfig().setLocation(i, sp.getLocation());
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "ハンター集合地点の位置" + i + "を設定しました。");
                                return;
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        } catch (NumberFormatException e) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        }
                    }
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location " + args[0] + " <数値>\" - ハンター集合地点の設定");
                    return;
                } else if (args[0].equalsIgnoreCase("door")) {
                    if (args.length != 1) {
                        if (args[1].equalsIgnoreCase("open")) {
                            if (args.length != 2) {
                                if (args[2].equalsIgnoreCase("all")) {
                                    worldConfig.getHunterDoorConfig().openHunterDoors();
                                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "設定したハンターボックスのドアすべてを開きました。");
                                    return;
                                } else {
                                    try {
                                        int i = Integer.parseInt(args[2]);
                                        if (i >= 1) {
                                            if (worldConfig.getConfig().contains(WorldManager.PathType.DOOR_HUNTER.getPath() + ".p" + i)) {
                                                worldConfig.getHunterDoorConfig().openHunterDoor(i);
                                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "ハンターボックスのドア" + i + "を開きました。");
                                                return;
                                            }
                                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "ハンターボックスのドア" + i + "は設定されていないため開くことができません。");
                                            return;
                                        }
                                        sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                                        return;
                                    } catch (NumberFormatException e) {
                                        sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                                        return;
                                    }
                                }
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door open <数値>\" - 指定したドアを開く\n" +
                                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door open all\" - 設定したすべてのドアを開く");
                            return;
                        } else {
                            try {
                                int i = Integer.parseInt(args[1]);
                                if (i >= 1) {
                                    worldConfig.getHunterDoorConfig().setLocation(i, sp.getLocation());
                                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "ハンターボックスのドア位置" + i + "を設定しました。");
                                    return;
                                }
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                                return;
                            } catch (NumberFormatException e) {
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                                return;
                            }
                        }
                    }
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door <数値>\" - ドア位置の設定\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door open <引数>\" - ドアを開く");
                    return;
                } else if (args[0].equalsIgnoreCase("jail")) {
                    if (args.length != 1) {
                        try {
                            int i = Integer.parseInt(args[1]);
                            if (i >= 1) {
                                worldConfig.getJailLocationConfig().setLocation(i, sp.getLocation());
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "牢獄地点の位置" + i + "を設定しました。");
                                return;
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        } catch (NumberFormatException e) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        }
                    }
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location jail <数値>\" - 牢獄地点の設定");
                    return;
                } else if (args[0].equalsIgnoreCase("respawn") || args[0].equalsIgnoreCase("res")) {
                    if (args.length != 1) {
                        try {
                            int i = Integer.parseInt(args[1]);
                            if (i >= 1) {
                                worldConfig.getRespawnLocationConfig().setLocation(i, sp.getLocation());
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "復活地点の位置" + i + "を設定しました。");
                                return;
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        } catch (NumberFormatException e) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。1以上で数字を指定してください。");
                            return;
                        }
                    }
                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                            MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location " + args[0] + " <数値>\" - 復活地点の設定");
                    return;
                }
                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location opgame\" または \"/location opg\" - オープニングゲーム地点の設定\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location gopgame <数値>\" または \"/location gopg <数値>\" - オープニングゲーム集合地点の設定\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location hunter\" - ハンター集合地点の設定\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door <引数…>\" - ドア位置の設定\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location jail <数値>\" - 牢獄地点の設定\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location respawn <数値>\" または \"/location res <数値>\" - 復活地点の設定");
                return;
            }
            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location opgame\" または \"/location opg\" - オープニングゲーム地点の設定\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location gopgame <数値>\" または \"/location gopg <数値>\" - オープニングゲーム集合地点の設定\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location hunter\" - ハンター集合地点の設定\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location door <引数…>\" - ドア位置の設定\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location jail <数値>\" - 牢獄地点の設定\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/location respawn <数値>\" または \"/location res <数値>\" - 復活地点の設定");
            return;
        }
        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
        return;
    }

    @Override
    public void onBlockCommand(BlockCommandSender bs, Command cmd, String label, String[] args) {
        MainAPI.sendMessage(bs, MainAPI.ErrorMessage.NOT_PLAYER);
        return;
    }

    @Override
    public void onConsoleCommand(ConsoleCommandSender cs, Command cmd, String label, String[] args) {
        MainAPI.sendMessage(cs, MainAPI.ErrorMessage.NOT_PLAYER);
        return;
    }

    @Override
    public List<String> onPlayerTabComplete(Player sp, Command cmd, String alias, String[] args) {
        if (TosoGameAPI.isAdmin(sp)) {
            if (args.length == 1) {
                return getTabList(args[0], new HashSet<>(Arrays.asList("opgame", "opg", "gopgame", "gopg", "hunter", "door", "jail", "respawn", "res")));
            } else if (args.length == 2) {
                if (args[0].equalsIgnoreCase("door")) {
                    return getTabList(args[1], new HashSet<>(Collections.singletonList("open")));
                }
            } else if (args.length == 3) {
                if (args[0].equalsIgnoreCase("door")) {
                    if (args[1].equalsIgnoreCase("open")) {
                        WorldConfig worldConfig = Main.getWorldConfig();

                        Set<String> set = new HashSet<>(Collections.singletonList("all"));
                        for (int i : worldConfig.getHunterDoorConfig().getLocations().keySet())
                            set.add(String.valueOf(i));
                        return getTabList(args[2], new HashSet<>(set));
                    }
                }
            }
        }
        return null;
    }

    @Override
    public List<String> onBlockTabComplete(BlockCommandSender bs, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onConsoleTabComplete(ConsoleCommandSender cs, Command cmd, String alias, String[] args) {
        return null;
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Location.kt
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.ErrorMessage
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Utils.ParseUtil
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.Command
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player

class Location(name: String) : ICommand(name) {

    override fun onPlayerCommand(sp: Player, cmd: Command, label: String, args: Array<String>) {
        if (TosoGameAPI.isAdmin(sp)) {
            val worldConfig = Main.worldConfig
            if (args.isNotEmpty()) {
                if (args[0].equals("opgame", true) || args[0].equals("opg", true)) {
                    worldConfig.opGameLocationConfig.opLocation = sp.location
                    sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}オープニングゲーム地点を設定しました。")
                    return
                } else if (args[0].equals("gopgame", true) || args[0].equals("gopg", true)) {
                    if (args.size != 1) {
                        if (ParseUtil.isInt(args[1])) {
                            val i = args[1].toInt()
                            if (i >= 1) {
                                worldConfig.opGameLocationConfig.setGOPLocation(i, sp.location)
                                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}オープニングゲーム集合地点の位置" + i + "を設定しました。")
                                return
                            }
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。1以上で数字を指定してください。")
                            return
                        } else {
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                            return
                        }
                    }
                    sp.sendMessage("""
                        ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                        ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location ${args[0]} <数値>" - オープニングゲーム集合地点の設定
                    """.trimIndent())
                    return
                } else if (args[0].equals("hunter", true)) {
                    if (args.size != 1) {
                        if (ParseUtil.isInt(args[1])) {
                            val i = args[1].toInt()
                            if (i >= 1) {
                                worldConfig.hunterLocationConfig.setLocation(i, sp.location)
                                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}ハンター集合地点の位置" + i + "を設定しました。")
                                return
                            }
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。1以上で数字を指定してください。")
                            return
                        } else {
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                            return
                        }
                    }
                    sp.sendMessage("""
                        ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                        ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location ${args[0]} <数値>" - ハンター集合地点の設定
                    """.trimIndent())
                    return
                } else if (args[0].equals("door", true)) {
                    if (args.size != 1) {
                        if (args[1].equals("all", true)) {
                            worldConfig.hunterDoorConfig.openHunterDoors()
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}設定したハンターボックスのドアすべてを開きました。")
                            return
                        } else {
                            if (ParseUtil.isInt(args[1])) {
                                val i = args[1].toInt()
                                if (i >= 1) {
                                    if (worldConfig.config.contains(WorldManager.PathType.DOOR_HUNTER.path + ".p" + i)) {
                                        worldConfig.hunterDoorConfig.openHunterDoor(i)
                                        sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}ハンターボックスのドア" + i + "を開きました。")
                                        return
                                    }
                                    sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}ハンターボックスのドア" + i + "は設定されていないため開くことができません。")
                                    return
                                }
                                sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。1以上で数字を指定してください。")
                                return
                            } else {
                                MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                                return
                            }
                        }
                    }
                    sp.sendMessage("""
                        ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                        ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location door open <数値>" - 指定したドアを開く
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location door open all" - 設定したすべてのドアを開く
                    """.trimIndent())
                    return
                } else if (args[0].equals("jail", true)) {
                    if (args.size != 1) {
                        if (ParseUtil.isInt(args[1])) {
                            val i = args[1].toInt()
                            if (i >= 1) {
                                worldConfig.jailLocationConfig.setLocation(i, sp.location)
                                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}牢獄地点の位置" + i + "を設定しました。")
                                return
                            }
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。1以上で数字を指定してください。")
                            return
                        } else {
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                            return
                        }
                    }
                    sp.sendMessage("""
                        ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                        ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location jail <数値>" - 牢獄地点の設定
                    """.trimIndent())
                    return
                } else if (args[0].equals("respawn", true) || args[0].equals("res", true)) {
                    if (args.size != 1) {
                        if (ParseUtil.isInt(args[1])) {
                            val i = args[1].toInt()
                            if (i >= 1) {
                                worldConfig.respawnLocationConfig.setLocation(i, sp.location)
                                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}復活地点の位置" + i + "を設定しました。")
                                return
                            }
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。1以上で数字を指定してください。")
                            return
                        } else {
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                            return
                        }
                    }
                    sp.sendMessage("""
                        ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                        ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                        ${MainAPI.getPrefix(PrefixType.ERROR)}"/location ${args[0]} <数値>" - 復活地点の設定
                    """.trimIndent())
                    return
                }
                sp.sendMessage("""
                    ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                    ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location opgame" または "/location opg" - オープニングゲーム地点の設定
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location gopgame <数値>" または "/location gopg <数値>" - オープニングゲーム集合地点の設定
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location hunter" - ハンター集合地点の設定
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location door <引数…>" - ドア位置の設定
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location jail <数値>" - 牢獄地点の設定
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/location respawn <数値>" または "/location res <数値>" - 復活地点の設定
                """.trimIndent())
                return
            }
            sp.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location opgame" または "/location opg" - オープニングゲーム地点の設定
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location gopgame <数値>" または "/location gopg <数値>" - オープニングゲーム集合地点の設定
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location hunter" - ハンター集合地点の設定
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location door <引数…>" - ドア位置の設定
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location jail <数値>" - 牢獄地点の設定
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/location respawn <数値>" または "/location res <数値>" - 復活地点の設定
            """.trimIndent())
            return
        }
        MainAPI.sendMessage(sp, ErrorMessage.PERMISSIONS)
        return
    }

    override fun onBlockCommand(bs: BlockCommandSender, cmd: Command, label: String, args: Array<String>) {
        MainAPI.sendMessage(bs, ErrorMessage.NOT_PLAYER)
        return
    }

    override fun onConsoleCommand(cs: ConsoleCommandSender, cmd: Command, label: String, args: Array<String>) {
        MainAPI.sendMessage(cs, ErrorMessage.NOT_PLAYER)
        return
    }

    override fun onPlayerTabComplete(sp: Player, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (TosoGameAPI.isAdmin(sp)) {
            if (args.size == 1) {
                return getTabList(args[0], "opgame", "opg", "gopgame", "gopg", "hunter", "door", "jail", "respawn", "res")
            } else if (args.size == 2) {
                if (args[0].equals("door", true)) {
                    val worldConfig = Main.worldConfig
                    val set = mutableSetOf("all")
                    for (i in worldConfig.hunterDoorConfig.doors.keys)
                        set.add(i.toString())
                    return getTabList(args[1], set)
                }
            }
        }
        return null
    }

    override fun onBlockTabComplete(bs: BlockCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }

    override fun onConsoleTabComplete(cs: ConsoleCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }
}